google/cloud/pubsublite/internal/multipartition_publisher.cc

```diff
@@ -169,24 +164,46 @@ void MultipartitionPublisher::RouteAndPublish(PublishState state) {
       .then(OnPublish{std::move(state.publish_promise), partition});
 }
 
+void MultipartitionPublisher::PublishLoop() {
+  bool messages_left;
+  {
+    std::lock_guard<std::mutex> g{mu_};
+    if (in_publish_loop_) return;
+    messages_left = !messages_.empty();
+    if (messages_left) in_publish_loop_ = true;
+  }
+  while (messages_left) {
+    PublishState state;
+    {
+      std::lock_guard<std::mutex> g{mu_};
+      state = std::move(messages_.front());
```

Comment:
Is there some guarantee that `messages_` is not empty at this point, or should there be a check right before calling `.front()` (under the lock) that it's still not empty to avoid the UB of calling front on a potentially empty deque?

Reply:
by using the `in_publish_loop_` flag variable, only one thread can execute `PublishLoop` at a time, and all other threads can only push to `messages_`
so our non-empty check before the loop and at the end of each iteration guarantee that it's non-empty when we call `front()` on it

Reply:
also restructured the code as per Dan's comments below, so we don't have to worry about this UB